src/main/java/com/epamlearning/services/impl/TrainingServiceImpl.java

```diff
@@ -0,0 +1,224 @@
+import com.epamlearning.microservices.report.dtos.TrainerWorkloadDTO;
+import com.epamlearning.repositories.TrainingRepository;
+import com.epamlearning.services.TrainingService;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.math.BigDecimal;
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+
+@Service
+@Slf4j
+public class TrainingServiceImpl implements TrainingService {
+
+  private final TrainingMapper trainingMapper;
+  private final TrainingRepository trainingRepository;
+  private final TraineeServiceImpl traineeService;
+  private final TrainerServiceImpl trainerService;
+
+  @Autowired
+  public TrainingServiceImpl(TrainingMapper trainingMapper, TrainingRepository trainingRepository,
+      TraineeServiceImpl traineeService, TrainerServiceImpl trainerService) {
+    this.trainingMapper = trainingMapper;
+    this.trainingRepository = trainingRepository;
+    this.traineeService = traineeService;
+    this.trainerService = trainerService;
+  }
+
+  public Training findById(Long id) {
+
+    if (id == null) {
+      log.warn("Training ID is null.");
+      throw new NotFoundException("Training ID is null.");
+    }
+
+    Optional<Training> training = trainingRepository.findById(id);
+    if (training.isEmpty()) {
+      log.warn("Training with ID: {} not found.", id);
+      throw new NotFoundException("Training with ID " + id + " not found.");
+    }
+    return training.get();
+  }
+
+  public boolean notExistsTraineeAndTrainerInAnotherTraining(Long trainingId,
+      String traineeUsername, String trainerUsername) {
+    Training training = findById(trainingId);
+    Trainee trainee = traineeService.findByValidatedUsername(traineeUsername);
+    Trainer trainer = trainerService.findByValidatedUsername(trainerUsername);
+    return !trainingRepository.existsAnotherTrainingByTraineeAndTrainer(training, trainee, trainer);
+  }
+
+  public boolean existsTraineeAndTrainerInTrainings(String traineeUsername,
+      String trainerUsername) {
+    Trainee trainee = traineeService.findByValidatedUsername(traineeUsername);
+    Trainer trainer = trainerService.findByValidatedUsername(trainerUsername);
+
+    return trainingRepository.existsTrainingByTraineeAndTrainer(trainee, trainer);
+  }
+
+
+  public Training update(Long id, Training training) {
+
+    if (training == null) {
+      log.warn("Training is null.");
+      throw new NotFoundException("Training is null.");
+    }
+    if (training.getTrainingDate() == null) {
+      log.warn("StartDate is null.");
+      throw new NotFoundException("StartDate is null.");
+    }
+    if (training.getTrainingDuration() == null) {
+      log.warn("TrainingDuration is null.");
+      throw new NotFoundException("TrainingDuration is null.");
+    }
```

Comment:
I would suggest moving this part of the code(**lines 80 - 91**) into a separate method for validation.